### PR TITLE
Add rosbag2_examples_cpp/simple_bag_reader.cpp.

### LINKED ISSUE #1449: Add additional examples for reading ROSBag data and serialized messages
## Description

Currently, the ROSBag2 examples contain three different examples for writing data to a ROSBag via C++  or Python. However, there are no specific examples of reading data from ROSBags (neither for the standard reader or the sequential reader). In addition, the writing examples do not show how to serialize various messages before writing them to a bag file.
Inside the file under _rosbag2_tests/test/rosbag2_tests/test_rosbag2_cpp_api.cpp_, there are test cases for reading data and also for the message serialization. These helped me when using the ROSBag2 API, but in my opinion, a better idea is to create dedicated examples, because the unit tests should be meant solely for testing.

## Completion Criteria

- Example for reading ROSBag data using a simple reader (`rosbag2_cpp::Reader`)
- Example for reading ROSBag data using a sequential reader (`rosbag2_cpp::SequentialReader`)
- Example for writing ROSBag data using serialized messages (`rclcpp::SerializedMessage` and `rclcpp::Serialization<>`)

### PR DESCRIPTION
**Pull request by fujitatomoya**
- Replaces https://github.com/ros2/rosbag2/pull/1580
 - Closes https://github.com/ros2/rosbag2/issues/1449

All other examples exist, and explained in ros2 documentation.

- data_generator_executable: https://docs.ros.org/en/rolling/Tutorials/Advanced/Recording-A-Bag-From-Your-Own-Node-CPP.html#write-a-c-executable
- data_generator_node: https://docs.ros.org/en/rolling/Tutorials/Advanced/Recording-A-Bag-From-Your-Own-Node-CPP.html#record-synthetic-data-from-a-node
- simple_bag_recorder: https://docs.ros.org/en/rolling/Tutorials/Advanced/Recording-A-Bag-From-Your-Own-Node-CPP.html#write-the-c-node

but this one is missing as a counterpart of https://docs.ros.org/en/rolling/Tutorials/Advanced/Reading-From-A-Bag-File-CPP.html#

related to https://github.com/ros2/ros2_documentation/pull/4499

as a follow-up for https://github.com/ros2/ros2_documentation, we need to make sure those examples are aligned with documentation. (`Reading from a bag file (Python)` needs to be added in doc.)

### CODE REVIEW

**Comment by fujitatomoya** [comment]
@MichaelOrlov can you take a look at this? this aligns with https://github.com/ros2/ros2_documentation/pull/4499 as executable sample.

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_reader.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  {
+    publisher_ = this->create_publisher<example_interfaces::msg::String>("chatter", 10);
+    timer_ = this->create_wall_timer(100ms,
+        [this](){return this->timer_callback();}
+    );
+
+    rosbag2_storage::StorageOptions storage_options;
+    storage_options.uri = bag_filename;
+    reader_ = rosbag2_transport::ReaderWriterFactory::make_reader(storage_options);
+    reader_->open(storage_options);
```
Comment (MichaelOrlov) [inline rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_reader.cpp:46]: To avoid situations where a timer callback is called but the reader hasn't been initialized yet, we need to init the reader first in the constructor.

**Review by MichaelOrlov — CHANGES_REQUESTED** [review]
@fujitatomoya Thank you for the PR. 
I think that it makes more sense to show users in examples how to use GenericPublisher without the deserialization of the recorded messages.
```
        std::shared_ptr<rclcpp::GenericPublisher> pub =
          owner_->create_generic_publisher(topic.name, topic.type, topic_qos);
```
https://github.com/ros2/rosbag2/blob/fbb08d04d2c07ac48187b58c9bec66a33cd0cd5f/rosbag2_transport/src/rosbag2_transport/player.cpp#L1137C1-L1138C79

**Comment by MichaelOrlov** [comment]
@fujitatomoya It looks like this PR very similar to the https://github.com/ros2/rosbag2/pull/1580/files which is already more thorough.
However, the https://github.com/ros2/rosbag2/pull/1580 is in a stale stage and need some fixes per code review. 
It would be great to cherry-pick commits from the https://github.com/ros2/rosbag2/pull/1580 and address review comments.

**Comment by fujitatomoya** [comment]
@MichaelOrlov thanks for pointing out https://github.com/ros2/rosbag2/pull/1580, that include python sample. i will try to cherry-pick some from there.

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_reader.cpp (hunk 1 above)
Comment (fujitatomoya) [reply rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_reader.cpp:46]: can you rephrase this? timer callback will never be called yet here until node starts spinning on the executor.

**Comment by fujitatomoya** [comment]
> I think that it makes more sense to show users in examples how to use GenericPublisher without the deserialization of the recorded messages.

this makes sense to me too.

File: rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_reader.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        sbr = SimpleBagReader(sys.argv[1])
+        rclpy.spin(sbr)
+    except KeyboardInterrupt:
+        pass
+    except ExternalShutdownException:
+        rclpy.shutdown()
+
+
+if __name__ == '__main__':
+    main()
```
Comment (fujitatomoya) [inline rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_reader.py]: we need final line feed here.

File: rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_recorder.py [hunk 3]
```diff
@@ -53,9 +54,13 @@ def topic_callback(self, msg):
...
-    sbr = SimpleBagRecorder()
-    rclpy.spin(sbr)
-    rclpy.shutdown()
+    try:
+        sbr = SimpleBagRecorder()
+        rclpy.spin(sbr)
+    except KeyboardInterrupt:
+        pass
+    except ExternalShutdownException:
+        rclpy.shutdown()
```
Comment (fujitatomoya) [inline rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_recorder.py:63]: this is required to avoid the following trace stack with `Ctrl-C` to stop the executable.

```bash
Traceback (most recent call last):
  File "/root/ros2_ws/colcon_ws/install/rosbag2_examples_py/lib/rosbag2_examples_py/simple_bag_reader", line 33, in <module>
    sys.exit(load_entry_point('rosbag2-examples-py', 'console_scripts', 'simple_bag_reader')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/ros2_ws/colcon_ws/build/rosbag2_examples_py/rosbag2_examples_py/simple_bag_reader.py", line 49, in main
    rclpy.spin(sbr)
  File "/root/ros2_ws/colcon_ws/install/rclpy/lib/python3.12/site-packages/rclpy/__init__.py", line 240, in spin
    executor.spin_once()
  File "/root/ros2_ws/colcon_ws/install/rclpy/lib/python3.12/site-packages/rclpy/executors.py", line 852, in spin_once
    self._spin_once_impl(timeout_sec)
  File "/root/ros2_ws/colcon_ws/install/rclpy/lib/python3.12/site-packages/rclpy/executors.py", line 836, in _spin_once_impl
    handler, entity, node = self.wait_for_ready_callbacks(
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/ros2_ws/colcon_ws/install/rclpy/lib/python3.12/site-packages/rclpy/executors.py", line 806, in wait_for_ready_callbacks
    return next(self._cb_iter)
           ^^^^^^^^^^^^^^^^^^^
  File "/root/ros2_ws/colcon_ws/install/rclpy/lib/python3.12/site-packages/rclpy/executors.py", line 710, in _wait_for_ready_callbacks
    wait_set.wait(timeout_nsec)
```

File: rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_reader.py (hunk 2 above)
Comment (fujitatomoya) [reply rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_reader.py]: addressed by https://github.com/ros2/rosbag2/pull/1683/commits/07a34eb24cec32b10ae7f60325db76c722c7877e

**Comment by fujitatomoya** [comment]
@MichaelOrlov requesting review.

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_reader.cpp (hunk 1 above)
Comment (MichaelOrlov) [reply rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_reader.cpp:46]: Oh, you are right. I forgot about it.

File: rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_recorder.py (hunk 3 above)
Comment (MichaelOrlov) [reply rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_recorder.py:63]: @fujitatomoya I'm wondering if it makes sense to add the section `finally` with `rclpy.shutdown()` as well?

File: rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_reader.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+
+class SimpleBagReader(Node):
+
+    def __init__(self, bag_filename):
+        super().__init__('simple_bag_reader')
+        self.reader = rosbag2_py.SequentialReader()
+        storage_options = rosbag2_py._storage.StorageOptions(
+            uri=bag_filename,
+            storage_id='sqlite3')
```
Comment (MichaelOrlov) [inline rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_reader.py:30]: @fujitatomoya I've double-checked, we actually don't use `storage_id` anywhere in the `SequantialReader` class. It would be better to show for users that it could be empty and don't need to bother to figure out what it is.

**Review by MichaelOrlov — COMMENTED** [review]
@fujitatomoya Thanks for the PR. 
Overall looks good to me among one nitpick with `storage_id` parameter and linter warning
```
uild/rosbag2_examples_py/pytest.xml: 3 tests, 0 errors, 1 failure, 0 skipped
2024-06-25T05:56:16.8541821Z - rosbag2_examples_py.test.test_flake8 test_flake8
2024-06-25T05:56:16.8542416Z   <<< failure message
2024-06-25T05:56:16.8542902Z     AssertionError: Found 1 code style errors / warnings:
2024-06-25T05:56:16.8543915Z       ./rosbag2_examples_py/simple_bag_reader.py:43:36: Q000 Double quotes found but single quotes preferred
2024-06-25T05:56:16.8544803Z     assert 1 == 0
2024-06-25T05:56:16.8545125Z   >>>
```

File: rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_reader.py (hunk 4 above)
Comment (fujitatomoya) [reply rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_reader.py:30]: @MichaelOrlov the thing is all existed examples in `rosbag2_examples_py` use `storage_id='sqlite3'`, so i just keep this aligned with those to make it work.

File: rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_recorder.py (hunk 3 above)
Comment (fujitatomoya) [reply rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_recorder.py:63]: thanks for the comment. if we use `finally` with `rclpy.shutdown()`, the following exception will be generated, because rclpy has been gracefully shut down by sigint. besides https://github.com/ros2/examples/tree/rolling/rclpy does not have `finally` either, so maybe we can keep this as it is.

```bash
rclpy._rclpy_pybind11.RCLError: failed to shutdown: rcl_shutdown already called on the given context, at /root/ros2_ws/colcon_ws/src/ros2/rcl/rcl/src/rcl/init.c:333
```

**Comment by fujitatomoya** [comment]
> @fujitatomoya Thanks for the PR. Overall looks good to me among one nitpick with `storage_id` parameter and linter warning
> 
> ```
> uild/rosbag2_examples_py/pytest.xml: 3 tests, 0 errors, 1 failure, 0 skipped
> 2024-06-25T05:56:16.8541821Z - rosbag2_examples_py.test.test_flake8 test_flake8
> 2024-06-25T05:56:16.8542416Z   <<< failure message
> 2024-06-25T05:56:16.8542902Z     AssertionError: Found 1 code style errors / warnings:
> 2024-06-25T05:56:16.8543915Z       ./rosbag2_examples_py/simple_bag_reader.py:43:36: Q000 Double quotes found but single quotes preferred
> 2024-06-25T05:56:16.8544803Z     assert 1 == 0
> 2024-06-25T05:56:16.8545125Z   >>>
> ```

addressed with https://github.com/ros2/rosbag2/pull/1683/commits/f1ba0f719580d50d49590a577e13818b14d5807b

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21328/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15641/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/838/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22059/)

**Comment by fujitatomoya** [comment]
@MichaelOrlov all green, requesting final review.

**Review by MichaelOrlov — APPROVED** [review]
